Add docs for artifact proxy (#5320)
mlflow/mlflow: docs/source/tracking.rst

MLflow entities (runs, parameters, metrics, tags, notes, metadata, etc), the artifact store persists artifacts
(files, models, images, in-memory objects, or model summary, etc).

> The MLflow client can be configured with an HTTP proxy, passing artifact requests through the tracking server to store and retrieve artifacts without having to specify the fully qualified path to the artifacts.

Reviewer: This is an MLflow server configuration, not a client configuration.

Author: yikes yeah. fixed.

`S3ArtifactRepository <https://github.com/mlflow/mlflow/blob/master/mlflow/store/artifact/s3_artifact_repo.py#L14>`_ are
concrete implementations of the abstract class `ArtifactRepository <https://github.com/mlflow/mlflow/blob/master/mlflow/store/artifact/artifact_repo.py#L13>`_.

Scenario 5: MLflow Tracking Server enabled as proxy for artifact storage access
-------------------------------------------------------------------------------

MLflow's Tracking Server supports utilizing the host as a proxy server for operations involving artifacts.
Once configured with the appropriate access requirements, an administrator can start the tracking server to enable assumed-role operations involving the saving, loading, or listing of
model artifacts, images, documents, and files. This eliminates the need to allow end users to have direct path access to a remote file store for artifact handling and eliminates the

> need for an end-user to provide access credentials to interact with an underlying file store.

Reviewer: Nit: "File store" is a bit of an overloaded term in MLflow, generally referring to the `FileStore` implementation of the backend metadata store API. Can we use "object store" instead?

Author: fixed and added references to object stores (just in case)

> .. figure:: _static/images/scenario_5.png

Reviewer: The "localhost" box in the `scenario_5` and `scenario_6` images show lines connecting from "RestStore" to "Remote Host. "RestStore" refers to the REST-based implementation of the backend metadata storage API, not an artifact storage API implementation. We should instead add a second small box referring to `HttpArtifactRepository` (https://github.com/mlflow/mlflow/blob/3c434246f73b41835194f3e6aa1f9be8488bcd56/mlflow/store/artifact/http_artifact_repo.py#L11).

Author: updated!

Enabling the Tracking Server to serve as an artifact proxy routes artifact operation requests and file transfers through the tracking server:

 * **Part 1a and b**:

  * The MLflow client creates an instance of a `RestStore` and sends REST API requests to log MLflow entities
  * The Tracking Server creates an instance of an `SQLAlchemyStore` and connects to the remote host for inserting
    tracking information in the database (i.e., metrics, parameters, tags, etc.)

 * **Part 1c and d**:

  * Retrieval requests by the client return information from the configured `SQLAlchemyStore` table

 * **Part 2a and b**:

  * Logging events for artifacts are made to the ``mlflow-artifacts:/`` uri for saving

>   * The Tracking server, serving as proxy, interfaces with the configured backend store, using its configured authorization to interact with the file store for writing artifacts

Reviewer: Specifically, the client uses `HttpArtifactRepository` (https://github.com/mlflow/mlflow/blob/3c434246f73b41835194f3e6aa1f9be8488bcd56/mlflow/store/artifact/http_artifact_repo.py#L11) to upload and download files to / from the remote host where the MLflow Tracking server is running.

Author: fixed the wording here.

  * The MLflow client creates an instance of a `RestStore` and sends REST API requests to log MLflow entities
  * The Tracking Server creates an instance of an `SQLAlchemyStore` and connects to the remote host for inserting
    tracking information in the database (i.e., metrics, parameters, tags, etc.)

 * **Part 1c and d**:

  * Retrieval requests by the client return information from the configured `SQLAlchemyStore` table

 * **Part 2a and b**:

  * Logging events for artifacts are made by the client using the ``HttpArtifactRepository`` to write files to MLflow Tracking Server
  * The Tracking Server then writes these files to the configured object store location with assumed role authentication

 * **Part 2c and d**:

  * Retrieving artifacts from the configured backend store for a user request is done with the same authorized authentication that was configured at server start
  * Artifacts are passed to the end user through the Tracking Server through the interface of the ``HttpArtifactRepository``

.. warning::
    The MLflow artifact proxied access service enables users to have an *assumed role of access to all artifacts* that are accessible to the Tracking Server.

>     Administrators who are enabling this feature should ensure that the access level granted to the Tracking Server for artifact operations meets with
>     any security requirements prior to enabling the Tracking Server to operate in a proxied file handling role.

Reviewer: ```suggestion
    Administrators who are enabling this feature should ensure that the access level granted to the Tracking Server for artifact operations meets all
    security requirements prior to enabling the Tracking Server to operate in a proxied file handling role.
```

.. _scenario_6:

Scenario 6: MLflow Tracking Server used exclusively as proxied access host for artifact storage access
------------------------------------------------------------------------------------------------------

> MLflow's Tracking Server can be used in an exclusive artifact proxied access file handling role. The configuration at server start by adding the flag
> ``--artifacts-only`` will restrict a Tracking Server instance to be used as a bastion host to an underlying file store without allowing for Tracking Server
> functionality apart from artifact handling. See: :ref:`artifact_only_mode` for more details.

Reviewer: "bastion host" is somewhat bespoke terminology. Can we simplify the phrasing here? 

```suggestion
MLflow's Tracking Server can be used in an exclusive proxied artifact handling role. Specifying the
``--artifacts-only`` flag restricts an MLflow server instance to only serve artifact-related API requests by
proxying to an underlying object store.
```

.. note::
    Starting a Tracking Server with the ``--artifacts-only`` parameter will disable all Tracking Server functionality apart from API calls related to saving, loading, or listing artifacts.
    Creating runs, logging metrics or parameters, and accessing other attributes about experiments are all not permitted in this mode.

.. figure:: _static/images/scenario_6.png

> Enabling the Tracking Server in ``--artifacts-only`` mode when enabling the instance to proxied artifact operations via ``--serve-artifacts`` disables other Tracking API functionality:

Reviewer: I think this is already stated in the `note` immediately above & can be removed.

        --default-artifact-root s3://my-mlflow-bucket/ \
        --host 0.0.0.0

An MLflow tracking server can also be run as a proxy artifact handler. An example configuration for the ``mlflow server`` in this mode is:

.. code-block:: bash

    mlflow server \
        --host 0.0.0.0 \
        --port 8889 \
        --serve-artifacts \

>         --default-artifact-root s3://my-mlflow-bucket/ \

Reviewer: Is this right? I think this should be `--artifacts-destination`.

        --default-artifact-root s3://my-mlflow-bucket/ \
        --host 0.0.0.0

> An MLflow tracking server can also be run as a proxied artifact handler. An example configuration for the ``mlflow server`` in this mode is:

Reviewer: ```suggestion
An MLflow Tracking server can also be run as a proxied artifact handler. An example configuration for the ``mlflow server`` in this mode is:
```

experiments that do not specify one. Once you create an experiment, ``--default-artifact-root``
is no longer relevant to that experiment.

> Starting a server with the ``--serve-artifacts`` option defined will enable proxy access for artifacts.

Reviewer: Nit: Instead of using future-tense "will" throughout the docs here, can we use present-tense phrasing (not an English expert, these are probably inaccurate tense names):

```suggestion
Starting a server with the ``--serve-artifacts`` option defined enables proxy access for artifacts.
```

One more small phrasing suggestion:

```suggestion
Starting a server with the ``--serve-artifacts`` option defined enables proxied access for artifacts.
```

Author: great call-out. Modified it here and a few other places.

> The uri ``mlflow-artifacts:/`` will serve in place of the ``--default-artifact-root`` path configured during
> server start to refer to artifact storage locations.

Reviewer: ```suggestion
The URI ``mlflow-artifacts:/`` is the default value for ``--default-artifact-root`` when `--serve-artifacts` is specified. This indicates to clients that artifacts can be accessed via HTTP requests to the MLflow Tracking Server.
```

Reviewer: Can we also enumerate and explain the use cases for some other proxied artifact `--default-artifact-root` values? E.g. `mlflow-artifacts://<tracking_server_host_name>/sub/path` or `https://<host_name>/sub/path`?

Author: Added a few listed examples - please check to make sure I'm capturing what you're looking for here or if I completely missed the plot.

.. important::
    When operating an MLflow tracking server with ``--serve-artifacts`` option enabled, the parameter
    ``--default-artifact-root`` will be set to ``mlflow-artifacts:/`` as a proxy root to the file store location

>     for artifact storage. Artifact handling can be accomplished through the use of this proxy uri. Access credentials

Reviewer: These first two sentences are redundant given the information above, right?

Author: definitely. removed them.

experiments that do not specify one. Once you create an experiment, ``--default-artifact-root``
is no longer relevant to that experiment.

> Starting a server with the ``--serve-artifacts`` option defined enables proxied access for artifacts.

Reviewer: ```suggestion
Starting a server with the ``--serve-artifacts`` flag enables proxied access for artifacts.
```

The uri ``mlflow-artifacts:/`` replaces an otherwise explicit object store destination (e.g., "s3:/my_bucket/mlartifacts")
for interfacing with artifacts. The client can access artifacts via HTTP requests to the MLflow Tracking Server.
This simplifies access requirements for users of the MLflow client, eliminating the need to
configure access tokens or username and password environment variables for the underlying object store when writing or retrieving artifacts.

Provided an Mlflow server configuraton where the ``--default-artifact-root`` is ``s3://my-root-bucket``,
the following patterns will all resolve to the configured proxied object store location of ``s3://my-root-bucket/mlartifacts``:

 * ``https://<host>:<port>/mlartifacts``
 * ``http://<host>/mlartifacts``
 * ``mlflow-artifacts://<host>/mlartifacts``
 * ``mlflow-artifacts://<host>:<port>/mlartifacts``

>  * ``mlflow-artifacts:/mlartifacts``

Reviewer: Can we clarify that, when the host is absent from the `mlflow-artifacts` URI, the client assumes that the host is the same as the host component of the tracking URI?

Additionally, you should ensure that the ``--backend-store-uri`` (which defaults to the
``./mlruns`` directory) points to a persistent (non-ephemeral) disk or database connection.

.. _artifact_only_mode:

> Using the Tracking Server exclusively for proxy artifact access

Reviewer: ```suggestion
Using the Tracking Server exclusively for proxied artifact access
```

Author: changed 17 references.

---------------------------------------------------------------

To use an instance of the MLflow tracking server *exclusively* for artifact operations ( :ref:`scenario_6` ),
start a server with the optional parameters ``--serve-artifacts`` to enable proxy artifact handling and ``--artifacts-only``
for disabling all other functionality of the tracking server.

To start the MLflow server in this restricted mode with a proxy to an HDFS location (as an example):

.. code-block:: bash

    mlflow server \
        --host 0.0.0.0 \
        --port 8885 \

>         --default-artifact-root hdfs://myhost:8887/mlprojects/models \

Reviewer: ```suggestion
        --artifacts-destination hdfs://myhost:8887/mlprojects/models \
```

Author: updated both of these references

        --serve-artifacts \
        --artifacts-only

Using this tracking server for any tasks aside from the aforementioned ``mlflow.<flavor>.log_model()``,

> ``mlflow.<flavor>.load_model()`` and ``client.list_artifacts(<run_id>)`` will throw an ``MLflowException``.

Reviewer: There are other methods used for logging artifacts, e.g. `mlflow.log_artifact()`. Perhaps we should update the phrasing to indicate that this isn't an exhaustive list of artifact-based APIs. This is also Python-specific.

Author: rephrased the entire paragraph.

Additionally, you should ensure that the ``--backend-store-uri`` (which defaults to the
``./mlruns`` directory) points to a persistent (non-ephemeral) disk or database connection.

.. _artifact_only_mode:

Using the Tracking Server exclusively for proxied artifact access
-----------------------------------------------------------------

> To use an instance of the MLflow tracking server *exclusively* for artifact operations ( :ref:`scenario_6` ),

Reviewer: ```suggestion
To use an instance of the MLflow Tracking server *exclusively* for artifact operations ( :ref:`scenario_6` ),
```

start a server with the optional parameters ``--serve-artifacts`` to enable proxied artifact access and ``--artifacts-only``

> for disabling all other functionality of the tracking server.

Reviewer: ```suggestion
for disabling all other functionality of the Tracking server.
```

To start the MLflow server in this restricted mode with proxied access to an HDFS location (as an example):

.. code-block:: bash

    mlflow server \
        --host 0.0.0.0 \
        --port 8885 \
        --artifacts-destination hdfs://myhost:8887/mlprojects/models \
        --serve-artifacts \
        --artifacts-only

> Using a tracking server configured in ``--artifacts-only`` mode for any tasks aside from those concerned with artifact

Reviewer: ```suggestion
Using an MLflow server configured in ``--artifacts-only`` mode for any tasks aside from those concerned with artifact
```

handling (i.e., model logging, loading models, logging artifacts, listing artifacts, etc.) will return an HTTPError.
See the following example of a client REST call in Python attempting to list experiments from a server that is configured in
``--artifacts-only`` mode:

.. code-block:: python

    import requests
    response = requests.get("http://0.0.0.0:8885/api/2.0/mlflow/experiments/list")

Output

.. code-block:: text

    >> HTTPError: Endpoint: /api/2.0/mlflow/experiments/list disabled due to the mlflow server running in `--artifacts-only` mode.

Using this mode to control access to artifacts without exposing the server endpoint's ability to create experiments,
manage runs, or perform any action apart from artifact handling can be useful in some scenarios (continuous deployment

> by an external team, for instance).

Reviewer: I think the biggest motivator is decoupling the compute / infrastructure used for serving these different types of requests, since each type of request (artifacts vs metadata) has different performance requirements and usage characteristics.
